backend/amd64/emit.ml

```diff
@@ -1362,6 +1361,193 @@ let emit_simd_instr op i =
 
+module Address_sanitizer : sig
+  type memory_access = Load | Store
+
+  (** Implements [https://github.com/google/sanitizers/wiki/AddressSanitizerAlgorithm#mapping]. *)
+  val emit_sanitize : ?dependencies:arg array -> address:arg -> memory_chunk -> memory_access -> unit
+
+  val command_line_options : (string * Arg.spec * string) list
+end = struct
+  type memory_access = Load | Store
+
+  module Memory_chunk_size = struct
+    type t =
+      | I8
+      | I16
+      | I32
+      | I64
+      | I128
+
+    let of_memory_chunk = function
+      | Byte_unsigned | Byte_signed -> I8
+      | Sixteen_unsigned | Sixteen_signed -> I16
+      | Thirtytwo_unsigned | Thirtytwo_signed | Single _ -> I32
+      | Word_int | Word_val | Double -> I64
+      | Onetwentyeight_unaligned | Onetwentyeight_aligned -> I128
+    ;;
+
+    external to_bytes_log2 : t -> int = "%identity"
+
+    let to_bytes t = 1 lsl to_bytes_log2 t
+  end
+
+  let mov_address src dest =
+    match src with
+    | Mem
+        { scale = 1;
+          base = None;
+          sym = None;
+          displ = 0;
+          idx;
+          arch = _;
+          typ = _
+        } ->
+      I.mov (Reg64 idx) dest
+    | _ -> I.lea src dest
+  ;;
+
+  let[@inline always] is_stack_16_byte_aligned () =
+    (* Yes, sadly this does result in materially better assembly than [(!stack_offset mod 16) = 0]
+       https://github.com/ocaml-flambda/flambda-backend/issues/2187 *)
+    !stack_offset land 15 = 0
+  ;;
+
+  let asan_report_function memory_chunk_size memory_access =
+    let index =
+      (Memory_chunk_size.to_bytes_log2 memory_chunk_size lsl 1)
+      + match memory_access with Load -> 0 | Store -> 1
+    in
+    (* We take extra care to structure our code such that these are statically
+       allocated as manifest constants in a flat array. *)
+    match index with
+    | 0 -> Sym "caml_asan_report_load1_noabort"
+    | 1 -> Sym "caml_asan_report_store1_noabort"
+    | 2 -> Sym "caml_asan_report_load2_noabort"
+    | 3 -> Sym "caml_asan_report_store2_noabort"
+    | 4 -> Sym "caml_asan_report_load4_noabort"
+    | 5 -> Sym "caml_asan_report_store4_noabort"
+    | 6 -> Sym "caml_asan_report_load8_noabort"
+    | 7 -> Sym "caml_asan_report_store8_noabort"
+    | 8 -> Sym "caml_asan_report_load16_noabort"
+    | 9 -> Sym "caml_asan_report_store16_noabort"
+    | _ ->
+      (* Larger loads and stores can be reported using
+         [__asan_report_load_n_noabort], but we don't support this yet. *)
+      assert false
+  ;;
+
+  let[@inline always] uses_register register = function
+    | Reg8L register' | Reg16 register' | Reg32 register' | Reg64 register' ->
+      register == register'
+    | Mem { idx = register'; base = None; scale; _ } ->
+      scale <> 0 && register == register'
+    | Mem { idx = register'; base = Some register''; _ } ->
+      register == register' || register == register''
+    | _ -> false
+  ;;
```

Comment:
Does it really make a difference to use phys equal here? Generally, I'm worried about the way `uses_register register` and `mov_address` break abstraction of X86_dsl. I don't know how to avoid it without complicating the interface of `Address_sanitizer`. @xclerc what do you suggest?

Reply:
A difference in terms of performance? No, probably not. But I didn't see an existing equality function for `reg64`. I could add one to `X86_ast` (which could still just be `phys_equal` under the hood if we want).

Reply:
Changed to use polymorphic equality.

Reply:
Discussed offline avoiding use of phys_equal here. Also, please add a CR about breaking the x86_dsl abstraction.

Reply:
Added.